reactions: Simplify code

showell opened the pull request:
I believe this will eliminate some strange bugs we've seen with duplicate emojis.  It should also make it easier to debug things down the road.

To tests this, have two users reacting to each other's messages.  Good things to try are editing messages and switching narrows from stream/topic to just-stream and back.

showell commented:
(still debugging a few things)

showell commented:
Ok, should be ready to review now.  Definitely QA this with the console open.

We can probably test-deploy this on czo for a couple hours if we want to err on the side of caution, since it's a significant rewrite.

timabbott commented:
I think the second commit is likely wrong; it appears to be removing the implementation for our handling of a rare type of corner case:
* A user reacts to a message
* We modify the mapping between names and emoji codes or the custom emoji they reacted with is disabled
* Another user clicks on the reaction to upvote it.  The desired UX is that the reaction gets upvoted rather than some sort of error happening.

But I'm not entirely sure as I haven't read this carefully; certainly that last commit appears to delete code and tests that look a lot like they're intended to handle this.

timabbott commented:
I think it'd also help me in reviewing this to have an understanding of what the model we're switching to is intended to be; I'm having a hard time reading the code changes, I think because I'm not intuiting what the structural shift is.

showell commented:
So the model here is really simple.  You create a Map called cleaned_reactions that is keyed on something like `unicode_emoji,1f3b1`, and then the map entries have things like `user_ids` and various description fields.

I don't have the energy to explain how awful the previous code, but the main problem was the data was completely denormalized:
- multiple buckets with the same key (all pointing to same list of user_ids)
- multiple buckets with the same emoji name

Most of the test code that I removed in the second commit was a test I wrote in the first commit. :)

And then I removed one test that was kinda contrived.

I urge you to just test-deploy this to czo for a day.

The code works very naturally for the common situation that an emoji code has multiple names.  We don't store the names in the key; we just make them attributes.  This is why you can eliminate the `get_alias_to_be_used` function.  The function was added in 60b0e0f8014aeef37f983cd701b674d25aa6e477 with no tests.

In the unlikely situation that a name actually migrated from one emoji code to another on some old message, I have zero confidence that the old code did the right thing, and what we'll do instead is report a blueslip error.

timabbott commented:
> In the unlikely situation that a name actually migrated from one emoji code to another on some old message, I have zero confidence that the old code did the right thing, and what we'll do instead is report a blueslip error.

The old code was specifically tested for this corner case when it was written, and I want to emphasize that this behaving correctly is an important feature; if we don't handle this correctly, the next time we upgrade emoji data, we'll get a bunch of complaints about reacting being broken.  So what I'm trying to say is that we can't have this become a blueslip error; it's something we should actually test.  

I can test-deploy this on chat.zulip.org for a day to see if it's obviously broken in other ways, which is probably a good idea, but we need to be confident that it handles the corner case of renamed emoji before we can merge this.

showell commented:
I think the miscommunication here (my fault for explaining poorly) is that changes happen in only one direction:

* an emoji with a certain code can have multiple names at the same time, can change names, can have deprecated names, etc., and my PR works fine with that
* an emoji name never migrates, as far as I know, from one code to another, and I don't believe the old code would have handled that correctly, and the symptom would be worse in the old code (whereas now it would be a clear blueslip error)

timabbott commented:
OK, I just hand-tested this PR with the exact corner case under discussion (after renaming the `rock_on` emoji to `wreck_on` manually), and it worked correctly.  So I think we're good on the detail :).  

I think it's totally possible for us to in the future have a name that used to point to one emoji code now point to another (though that will be dominated by someone uploading a realm emoji with a given name).  I tested the "upload realm emoji" corner case and determined it works for messages that don't already have a reaction from us using the old name; for those that do, we send a request and get a "Reaction already exists" error, which is probably an acceptable result.

showell commented:
This was test-deployed on czo about 20 hours ago, and no problems have been reported via either human users or blueslip errors.

I think a relative newbie with good JS skills could probably extract `reaction_data.js` now, which might be useful for mobile code sharing.  It would be nice to kind of isolate the UI layer here, since we may just wanna go with vdom eventually.
